### PR TITLE
Member map

### LINKED ISSUE #47: [member_map] implements selectKRandomMember
implements `selectKRandomMember`

**reference**
https://github.com/DE-labtory/swim/blob/develop/member_map.go

### PR DESCRIPTION
**Pull request by junbeomlee**
resolved: #47

details:

1. Implement select k random memberID and add member and shuffle function
2. Fix bug in `Push function` pbkstore.go

- [x] Test case

### CODE REVIEW

File: member_map.go [hunk 1]
```diff
@@ -17,66 +17,215 @@
...
+	}
+
+	// member2 == member1 case suspect status can override alive
+	if newMem.Incarnation == existingMem.Incarnation {
+		if newMem.Status == Suspected && existingMem.Status == Alive {
+			return newMem
+		}
+		if newMem.Status == Suspected && existingMem.Status == Alive {
+			return newMem
+		}
```
Comment (leesd556) [inline member_map.go]: Duplicate 'if' statement

Comment (AgwaB) [inline member_map.go]: Duplicate 'if' statement.

**Review by AgwaB — COMMENTED** [review]
I suggest `RemoveMember` function in `member_map.go`

```go
func (m *MemberMap) RemoveMember(member Member) error {
	m.lock.Lock()
	defer m.lock.Unlock()

	if member.GetID().ID == "" {
		return ErrEmptyMemberID
	}

	_, ok := m.members[member.GetID()]
	if ok {
		delete(m.members, member.GetID())
	}

	m.waitingMembers = resetWaitingMembersID(m.members)
	return nil
}
```

File: member_map.go [hunk 2]
```diff
@@ -17,66 +17,215 @@
...
+
+	m.waitingMembers = resetWaitingMembersID(m.members)
+	return nil
+}
+
+// Update member and update waitingMembers
+// todo
+func (m *MemberMap) UpdateMember(member Member) error {
+	return nil
+}
```
Comment (AgwaB) [inline member_map.go:167]: I think it is very similar to `AddMember`. What do you think?

File: member_map.go [hunk 3]
```diff
@@ -17,66 +17,215 @@
...
+		members:        make(map[MemberID]Member),
+		waitingMembers: make([]MemberID, 0),
+		lock:           sync.RWMutex{},
+	}
 }
 
-// select K random member from members
-func (m *MemberMap) selectKRandomMember(k int) []Member {
+// Select K random memberID from waitingMembers(length of returning member can be lower than k).
+// ** WaitingMembers are shuffled every time when members are updated **, so just returning first K item in waitingMembers is same as
```
Comment (zeroFruit) [inline member_map.go:104]: > ** WaitingMembers are shuffled every time when members are updated **

I think this comment will be more precise 

> ** WaitingMembers are shuffled every time when waiting members are updated **

I think there's no logic updating **waiting members** when **members** are updated

File: member_map.go (hunk 2 above)
Comment (junbeomlee) [reply member_map.go:167]: I agree. I should think about all cases, if `add and update function` do same things, then update and add function will be merged. So i marked this function(todo)

File: member_map.go (hunk 1 above)
Comment (junbeomlee) [reply member_map.go]: fixed it.

File: member_map_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+		ID: MemberID{
+			ID: "2",
+		},
+		Status:      Alive,
+		Incarnation: 1,
+	}
+
+	assert.Equal(t, override(m3, m4), m4)
+}
+
```
Comment (zeroFruit) [inline member_map_internal_test.go:243]: I think there's no test cases when 

* {Confirm Ml, inc=i} overrides
    - {Alive Ml, inc=j}, any j
    - {Alive Ml, inc=j}, any j

for **any j** cases

File: member_map.go (hunk 3 above)
Comment (junbeomlee) [reply member_map.go:104]: `m.waitingMembers = resetWaitingMembersID(m.members)` this function shuffles waiting members using members. `m.waitingMembers = resetWaitingMembersID(m.members)` located in the lower part of addmember function

**Comment by junbeomlee** [comment]
```
func (m *MemberMap) RemoveMember(member Member) error {
	m.lock.Lock()
	defer m.lock.Unlock()

	if member.GetID().ID == "" {
		return ErrEmptyMemberID
	}

	_, ok := m.members[member.GetID()]
	if ok {
		delete(m.members, member.GetID())
	}

	m.waitingMembers = resetWaitingMembersID(m.members)
	return nil
}
```
This function also need a test case so, could you implement this function after merge this PR.

File: member_map.go (hunk 3 above)
Comment (zeroFruit) [reply member_map.go:104]: yes `resetWatingMembersID(m.members)` shuffles waiting members using *members* but I think there's no **when members are updated** part in the function

**Comment by junbeomlee** [comment]
```
func (m *MemberMap) AddMember(member Member) error {
	m.lock.Lock()
	defer m.lock.Unlock()

	// If id is empty return error
	if member.GetID().ID == "" {
		return ErrEmptyMemberID
	}

	// Check whether it is already exist
	exMem, ok := m.members[member.GetID()]
	if ok {
		// Apply override Rule
		member = override(member, exMem)
	}

	m.members[member.GetID()] = member

	m.waitingMembers = resetWaitingMembersID(m.members)
	return nil
}
```
Add member function update members.
`** WaitingMembers are shuffled every time when waiting members are updated **` This is an additional description of when waiting members are updated.

**Comment by zeroFruit** [comment]
#55 updated

**Review by zeroFruit — APPROVED** [review]

**Pull request merged**